dev/archery/archery/cli.py

```diff
@@ -0,0 +1,304 @@
+@click.option("--debug", type=bool, is_flag=True, default=False,
+              help="Increase logging with debugging output.")
+@click.option("-q", "--quiet", type=bool, is_flag=True, default=False,
+              help="Silence executed commands.")
+@click.pass_context
+def archery(ctx, debug, quiet):
+    """ Apache Arrow developer utilities.
+
+    See sub-commands help with `archery <cmd> --help`.
+
+    """
+    # Ensure ctx.obj exists
+    ctx.ensure_object(dict)
+
+    log_ctx.quiet = quiet
+    if debug:
+        logger.setLevel(logging.DEBUG)
+
+
+def validate_arrow_sources(ctx, param, src):
+    """ Ensure a directory contains Arrow cpp sources. """
+    if isinstance(src, str):
+        if not ArrowSources.valid(src):
+            raise click.BadParameter(f"No Arrow C++ sources found in {src}.")
+        src = ArrowSources(src)
+    return src
+
+
+build_dir_type = click.Path(dir_okay=True, file_okay=False, resolve_path=True)
+# Supported build types
+build_type = click.Choice(["debug", "relwithdebinfo", "release"],
+                          case_sensitive=False)
+# Supported warn levels
+warn_level_type = click.Choice(["everything", "checkin", "production"],
+                               case_sensitive=False)
+
+
+@archery.command(short_help="Initialize an Arrow C++ build")
+@click.option("--src", metavar="<arrow_src>", default=ArrowSources.find(),
+              callback=validate_arrow_sources,
+              help="Specify Arrow source directory")
+# toolchain
+@click.option("--cc", metavar="<compiler>", help="C compiler.")
+@click.option("--cxx", metavar="<compiler>", help="C++ compiler.")
+@click.option("--cxx-flags", help="C++ compiler flags.")
+@click.option("--build-type", default="release", type=build_type,
+              help="CMake's CMAKE_BUILD_TYPE")
+@click.option("--warn-level", default="production", type=warn_level_type,
+              help="Controls compiler warnings -W(no-)error.")
+# components
+@click.option("--with-tests", default=True, type=bool,
+              help="Build with tests.")
+@click.option("--with-benchmarks", default=False, type=bool,
+              help="Build with benchmarks.")
+@click.option("--with-python", default=True, type=bool,
+              help="Build with python extension.")
+@click.option("--with-parquet", default=False, type=bool,
+              help="Build with parquet file support.")
+@click.option("--with-gandiva", default=False, type=bool,
+              help="Build with Gandiva expression compiler support.")
+@click.option("--with-plasma", default=False, type=bool,
+              help="Build with Plasma object store support.")
+@click.option("--with-flight", default=False, type=bool,
+              help="Build with Flight rpc support.")
+# misc
+@click.option("-f", "--force", type=bool, is_flag=True, default=False,
+              help="Delete existing build directory if found.")
+@click.option("--targets", type=str, multiple=True,
+              help="Generator targets to run")
```

Comment:
An option like `"--cmake-extra-args"` will probably be handy at some point, for example when tweaking `CMakeLists.txt`
```suggestion
              help="Generator targets to run")
@click.option("--cmake-extra-args", type=str, multiple=True,
              help="Extra arguments to pass to CMake")
```

Reply:
Absolutely, I don't know why I forgot it, I actually have it my shell script I derived this command from.

https://gist.github.com/fsaintjacques/b4302b2f8a218dadfad5a2b942cef2fc